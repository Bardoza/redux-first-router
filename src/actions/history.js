// @flow
import { CALL_HISTORY } from '../types'

export const push = (path: string, state: ?Object) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'push',
    args: [path, state]
  }
})

export const replace = (path: string, state: ?Object) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'replace',
    args: [path, state]
  }
})

export const jump = (n: number | string, state: ?Object, byIndex: ?boolean, kind: ?string) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'jump',
    args: [n, state, byIndex, kind]
  }
})

export const reset = (entries: Array<string | Object>, index: ?number, kind: ?string) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'reset',
    args: [entries, index, kind]
  }
})

export const setState = (state: Object | Function, n: ?(number | string), byIndex: ?boolean) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'setState',
    args: [state, n, byIndex]
  }
})

export const back = (state: ?(Object | Function)) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'back',
    args: [state]
  }
})

export const next = (state: ?(Object | Function)) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'next',
    args: [state]
  }
})


// NOTE: see `src/utils/formatRoutes.js` for implemenation of corresponding pathlessRoute
