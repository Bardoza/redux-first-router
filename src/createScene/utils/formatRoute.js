// @flow
import { isNotFound, typeToScene, formatRoute } from '../../utils'

export default (r, type, routes, formatter) => {
  const route = formatRoute(r, type, routes, formatter)

  route.scene = typeToScene(type)

  // set default path for NOT_FOUND actions if necessary
  if (!route.path && isNotFound(type)) {
    route.path = route.scene
      ? `/${r.scene.toLowerCase()}/not-found`
      : '/not-found'
  }

  return route
}
