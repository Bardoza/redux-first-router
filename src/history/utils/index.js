export {
  supportsDom,
  supportsHistory,
  supportsSessionStorage
} from './supports'

export {
  addPopListener,
  removePopListener,
  isExtraneousPopstateEvent
} from './popListener'

export {
  saveHistory,
  restoreHistory,
  getHistoryState,
  getInitialHistoryState,
  getInitialN
} from './sessionStorage'
