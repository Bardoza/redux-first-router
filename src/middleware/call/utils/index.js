export { default as enhanceRoutes } from './enhanceRoutes'
export { default as onError } from './onError'
export { default as shouldCall } from './shouldCall'
export { default as createCache } from './createCache'
export { default as autoDispatch } from './autoDispatch'
