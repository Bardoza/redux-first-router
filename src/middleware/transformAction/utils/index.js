export { default as formatAction } from './formatAction'
export { default as replacePopAction, findNeighboringN } from './replacePopAction'
export { default as handleDoubleDispatch, isDoubleDispatch } from './doubleDispatch'
export { default as createNotFoundRoute } from './createNotFoundRoute'
export { default as nestAction, createActionRef, createStateRef } from './nestAction'

