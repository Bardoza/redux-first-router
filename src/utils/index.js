export { default as isHydrate } from './isHydrate'
export { default as isAction } from './isAction'
export { default as isNotFound } from './isNotFound'
export { default as isServer } from './isServer'
export { default as isRedirect } from './isRedirect'

export { default as actionToUrl } from './actionToUrl'
export { default as urlToAction } from './urlToAction'

export { default as doesRedirect } from './doesRedirect'
export { default as shouldTransition } from './shouldTransition'

export { default as matchUrl } from './matchUrl'
export { default as compileUrl } from './compileUrl'

export { default as formatRoutes, formatRoute } from './formatRoutes'
export { default as typeToScene } from './typeToScene'

export { default as noOp } from './noOp'

export { default as createSelector } from './createSelector'
export { default as createAction } from './createAction'

export { default as redirectShortcut } from './redirectShortcut'
export { default as callRoute } from './callRoute'
