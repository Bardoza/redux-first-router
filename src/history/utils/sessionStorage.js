import { supportsSessionStorage } from './index'
import { urlToAction } from '../../utils'

// PREFIXING:

const PREFIX = '@@rudy/'

const prefixKey = key => PREFIX + getId() + key

// BASIC SESSIONSTORAGE WRAPPER METHODS:

const setItem = (key, val) => {
  window.sessionStorage.setItem(prefixKey(key), JSON.stringify(val))
}

export const getItem = key => {
  try {
    const json = window.sessionStorage.getItem(prefixKey(key))
    return JSON.parse(json)
  }
  catch (error) {
    // Ignore invalid JSON.
  }
}

const getSetItem = (key, val) => {
  const item = getItem(key)

  if (!item) {
    setItem(key, val)
  }

  return item || val
}


// PRIMARY FACADE METHODS:

// Below is the facade around both `sessionStorage` and our `history-storage` fallback.
// Essentially the idea is that if there is no `sessionStorage`, we maintain the entire
// storage object on EACH AND EVERY history entry's `state`. I.e. `history.state` on
// every page will have the `index` and `entries` array. That way when browsers disable
// cookies/sessionStorage, we can still grab the data we need off off of history state :)
//
// It's a bit crazy, but it works very well, and there's plenty of space allowed for storing
// things there to get a lot of mileage out of it. Firefox has the lowest limit of 640kb per
// entry. IE has 1mb and chrome has at least 10mb:
// https://stackoverflow.com/questions/6460377/html5-history-api-what-is-the-max-size-the-state-object-can-be

// called every time the history entries or index changes
export const saveHistory = ({ index, entries, forwardedOut }) => {
  entries = entries.map(e => [e.location.url, e.state, e.location.key])

  const history = forwardedOut
    ? { index, entries, forwardedOut } // used to patch an edge case, see `getIndexAndEntries` below
    : { index, entries }

  // here's the key aspect of the fallback. essentially we keep updating history state
  // via `replaceState` so every entry has everything that would be in `sessionStorage`
  if (!supportsSessionStorage()) {
    const state = getHistoryState()
    delete state.forwardedOut
    const newState = { ...state, history }

    window.history.replaceState(newState, null, window.location.href)
    return
  }

  setItem('history', history)
}

// called on startup
export const restoreHistory = (defaultLocation, api) => {
  const { state, location: { url, key } } = defaultLocation // used if first time visiting site during session
  const defaultHistory = { n: 1, index: 0, entries: [[url, state, key]] }

  if (!supportsSessionStorage()) {
    const state = getHistoryState()

    // we gotta of course call `replaceState` on first load as well
    // to insure this entry has all the state. The rest are handled by
    // `saveHistory` on navigation.
    if (!state.history) {
      const newState = { ...state, history: defaultHistory }
      window.history.replaceState(newState, null, window.location.href)
    }

    return getIndexAndEntries(state.history || defaultHistory, api)
  }

  // `getSet` simply sets the `defaultHistory` in `sessionStorage` if it's a fresh visitation
  return getIndexAndEntries(getSetItem('history', defaultHistory), api)
}

export const getHistoryState = () => {
  try {
    return window.history.state || {}
  }
  catch (e) {
    // IE 11 sometimes throws when accessing window.history.state
    // See: https://github.com/ReactTraining/history/pull/289
    // If you follow the links, you'll see the issue A) occurs when you refresh
    // a page that is the 1 and only entry in history, i.e. would never have any
    // state to remember in the first place; and B) is IE11 on load in iframes,
    // which also won't need to remember state, as iframes usually aren't for navigating
    // to other sites. IE11 should generally work fine when navigating to and from your site.
    return {}
  }
}

// this is a key one, as it sets the id in history.state for the first entry
// that way the `id` can be used to differentiate between different sessionStorages
// that could happen in the same tab if you manually enter another URL at your site.
// It's a bit complex, but basically we use an id from history state to discover
// multiple possible available sessionStorages lol. Yes, our ware is robust :)
export const getInitialHistoryState = () => {
  const currentState = getHistoryState()

  const id = (_id = currentState.id || createKey())
  const key = currentState.key || createKey()
  const newState = Object.assign({}, currentState, { id, key })
  window.history.replaceState(newState, null, window.location.href)

  return newState
}


// FACADE HELPERS:

const getIndexAndEntries = (history, api) => {
  let { index, entries, forwardedOut } = history

  // We must remove entries after the index in case the user opened a link to
  // another site in the middle of the entries stack and then returned via the
  // back button, in which case the entries are gone for good, like a `push`.
  // NOTE: we currently don't re-save the sliced array, as it will be saved
  // on any future change--because if the user, for example, refreshed, the
  // same thing would happen.
  //
  // EXCEPTION: if we did this on the first entry, we would break backing out of the
  // site and returning (entries would be unnecessarily removed). So this is only applied to
  // "forwarding out." That leaves one hole: if you forward out from the first entry, you will
  // return and have problematic entries that should NOT be there. Then because of Rudy's
  // automatic back/next detection, which causes the history track to "jump" instead of "push,"
  // dispatching an action for the next entry would in fact make you leave the site instead
  // of push the new entry! To circumvent that, use Rudy's <Link /> component and it will
  // set the `forwardedOut` flag (just before linking out) that insures this is addressed:
  if (index > 0 || forwardedOut) {
    entries = entries.slice(0, index + 1)
  }

  entries = entries.map(([url, state, key]) => urlToAction(url, api, state, key))
  const n = getInitialN(index, entries)
  return { n, index, entries }
}

let _id

const getId = () => {
  _id = _id || getHistoryState().id || 'id' // sessions when using memory history in the browser won't be able to have unique IDs
  return `${_id}/`
}

const createKey = () =>
  Math.random().toString(36).substr(2, 6)

// When entries are restored on load, the direction is always forward if on an index > 0
// because the corresponding entries are removed (just like a `push`), and you are now at the head.
//
// Otherwise, if there are multiple entries and you are on the first, you're considered
// to be going back, but if there is one, you're logically going forward.
export const getInitialN = (index, entries) =>
  index > 0 ? 1 : (entries.length > 1 ? -1 : 1)
