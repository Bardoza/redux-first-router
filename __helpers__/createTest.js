import { applyMiddleware, createStore, combineReducers } from 'redux'
import { createRouter } from '../src'

export default async (...allArgs) => {
  const args = allArgs.filter(arg => typeof arg !== 'function')
  const callbacks = allArgs.filter(arg => typeof arg === 'function')
  const hasCallbacks = callbacks.length > 0

  const [testName, routesMap] = args
  let [,, options = {}, actions] = args

  if (Array.isArray(options)) {
    actions = options
    options = {}
  }

  if (options.skip) return

  options.browser = options.browser || false

  if (!actions) {
    actions = Object.keys(routesMap).filter(type => !/FIRST|REDIRECTED/.test(type))
  }

  const initialPath = typeof actions[0] === 'string' && actions[0].charAt(0) === '/'
    ? actions.shift()
    : '/first'

  const hasMultipleTests = actions.length > 1
  let num = 1

  if (hasMultipleTests) {
    describe(testName, () => {
      for (const action of actions) {
        if (Array.isArray(action)) {
          const name = action[0]
          const act = action[1]
          createTest(name, routesMap, initialPath, act, options, num++)
        }
        else {
          const name = JSON.stringify(action)
          createTest(name, routesMap, initialPath, action, options, num++)
        }
      }

      if (hasCallbacks) {
        callbacks.forEach((cb, index) => {
          const name = callbacks.length === 1 ? 'snipes' : 'snipes ' + (++index)
          createSnipes(name, routesMap, initialPath, options, cb)
        })
      }
    })
  }
  else {
    const isSnipesOnly = initialPath === '/first' && actions.length === 0

    if (hasCallbacks && isSnipesOnly) {
      callbacks.forEach((cb, index) => {
        const name = callbacks.length === 1 ? testName : 'snipes ' + (++index)
        createSnipes(name, routesMap, initialPath, options, cb)
      })
    }
    else if (hasCallbacks) {
      describe(testName, () => {
        const name = actions[0] ? JSON.stringify(actions[0]) : 'firstRoute - ' + initialPath

        createTest(name, routesMap, initialPath, actions[0], options, num)

        callbacks.forEach((cb, index) => {
          const name = callbacks.length === 1 ? 'snipes' : 'snipes ' + (++index)
          createSnipes(name, routesMap, initialPath, options, cb)
        })
      })
    }
    else if (Array.isArray(actions[0])) {
      const name = actions[0][0]
      const act = actions[0][1]

      createTest(name, routesMap, initialPath, act, options, num)
    }
    else {
      createTest(testName, routesMap, initialPath, actions[0], options, num)
    }
  }
}

const createTest = (testName, routesMap, initialPath, item, opts, num) => {
  test(testName, async () => {
    const {
      store,
      history,
      routes,
      options,
      firstRoute,
      initialState
    } = setupStore(routesMap, initialPath, opts)

    const firstAction = firstRoute(false)
    const res = await store.dispatch(firstAction)

    if (routesMap.FIRST || initialPath !== '/first') {
      const prefix = 'firstRoute - ' + initialPath + ' - ' + num
      snapChange(prefix, res, store, history, initialState)
    }

    if (typeof item === 'string' && item.charAt(0) === '/') {
      const url = item
      const res = await history.push(url)

      snapChange(num, res, store, history)
    }
    else if (item) {
      const action = typeof item === 'string' ? { type: item } : item
      const res = await store.dispatch(action)

      snapChange(num, res, store, history)
    }

    snapRoutes(num, routes)
    snapOptions(num, options)

    if (opts.log) console.log(store.getState().location)
  })
}

const createSnipes = (testName, routesMap, initialPath, opts, callback) => {
  test(testName, async () => {
    const {
      store,
      history,
      routes,
      options,
      firstRoute
    } = setupStore(routesMap, initialPath, opts)

    const firstResponse = opts.dispatchFirstRoute === false
      ? null
      : await store.dispatch(firstRoute(false))

    const pop = opts.browser && createPop(history)

    await callback({
      firstRoute,
      firstResponse,
      history,
      routes,
      options,
      store,
      pop,
      dispatch: store.dispatch,
      getState: store.getState,
      getLocation: () => store.getState().location,
      snapChange,
      snap: async (action, prefix = '') => {
        const res = await store.dispatch(action)

        prefix = prefix || JSON.stringify(action) || 'function'

        snapChange(prefix, res, store, history)
        snapRoutes(prefix, routes)
        snapOptions(prefix, options)

        return res
      },
      snapPop: async (direction, prefix = '') => {
        const res = await pop(direction)

        snapChange(prefix, res, store, history)
        snapRoutes(prefix, routes)
        snapOptions(prefix, options)

        return res
      }
    })

    if (opts.log) console.log(store.getState().location)
  })
}

export const setupStore = (routesMap, initialPath, opts) => {
  const routes = createRoutes(routesMap)
  const options = createOptions(opts)

  options.initialEntries = [initialPath]
  options.extra = options.extra || { arg: 'extra-arg' }

  const middlewareFunc = options.middlewareFunc
  delete options.middlewareFunc

  const title = (state, action = {}) => {
    return action.payload !== undefined
      ? action.type + ' - ' + JSON.stringify(action.payload)
      : action.type
  }

  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options,
    middlewareFunc
  )

  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware)
  const store = createStore(rootReducer, enhancer)

  const initialState = store.getState()

  return {
    store,
    firstRoute,
    initialState,
    routes,
    options,
    history
  }
}

const callbacks = [
  'beforeLeave',
  'beforeEnter',
  'onLeave',
  'onLeave',
  'onEnter',
  'thunk',
  'onComplete',
  'onError'
]

const createRoutes = (routesMap) => {
  const routes = {}

  for (const type in routesMap) {
    routes[type] = typeof routesMap[type] === 'object'
      ? { ...routesMap[type] }
      : routesMap[type]

    const route = routes[type]

    for (const cb of callbacks) {
      if (typeof route[cb] === 'function') {
        route[cb] = jest.fn(route[cb])
      }
    }
  }

  return {
    FIRST: {
      path: '/first'
    },
    NEVER_USED_PATHLESS: { // insures pathless routes can co-exist with regular routes
      thunk: jest.fn()
    },
    REDIRECTED: {
      path: '/redirected',
      beforeEnter: () => {
        return new Promise(res => setTimeout(res, 1))
      },
      onComplete: jest.fn(() => 'redirect_complete')
    },
    ...routes
  }
}

const createOptions = (opts = {}) => {
  const options = { ...opts }

  for (const cb of callbacks) {
    if (typeof opts[cb] === 'function') {
      options[cb] = jest.fn(opts[cb])
    }
  }

  return options
}

const snapRoutes = (prefix, routes) => {
  for (const type in routes) {
    const route = routes[type]
    snapCallbacks(prefix + ' - routes - ' + type, route)
  }
}

const snapOptions = (prefix, options) => {
  snapCallbacks(prefix + ' - options', options)
}

const snapCallbacks = (prefix, obj) => {
  expectBeforeLeave(prefix, obj)
  expectBeforeEnter(prefix, obj)
  expectOnLeave(prefix, obj)
  expectOnEnter(prefix, obj)
  expectThunk(prefix, obj)
  expectOnComplete(prefix, obj)
  expectOnError(prefix, obj)
}

const snapChange = (prefix, res, store, history, initialState) => {
  if (initialState) expectInitialState(prefix, initialState)

  expectState(prefix, store)
  expectResponse(prefix, res)
  expectEntries(prefix, history)
  expectIndex(prefix, history)
  expectTitle(prefix)
}

// all these expect functions are broken out separately so we can easily see the
// name of the expectation that failed in the trace displayed in the Wallaby web panel

const expectInitialState = (prefix, initialState) => {
  expect(initialState).toMatchSnapshot(prefix + ' - initial_state')
}

const expectState = (prefix, store) => {
  expect(store.getState()).toMatchSnapshot(prefix + ' - state')
}

const expectResponse = (prefix, res) => {
  expect(res).toMatchSnapshot(prefix + ' - response')
}

const expectEntries = (prefix, history) => {
  expect(history.entries).toMatchSnapshot(prefix + ' - history_entries')
}

const expectIndex = (prefix, history) => {
  expect(history.index).toMatchSnapshot(prefix + ' - history_index')
}

const expectTitle = (prefix) => {
  expect(document.title).toMatchSnapshot(prefix + ' - title')
}

const expectBeforeLeave = (prefix, obj) => {
  if (typeof obj.beforeLeave === 'function' && obj.beforeLeave.mock) {
    expect(obj.beforeLeave.mock.calls.length).toMatchSnapshot(prefix + ' - beforeLeave')
  }
}

const expectBeforeEnter = (prefix, obj) => {
  if (typeof obj.beforeEnter === 'function' && obj.beforeEnter.mock) {
    expect(obj.beforeEnter.mock.calls.length).toMatchSnapshot(prefix + ' - beforeEnter')
  }
}

const expectOnLeave = (prefix, obj) => {
  if (typeof obj.onLeave === 'function' && obj.onLeave.mock) {
    expect(obj.onLeave.mock.calls.length).toMatchSnapshot(prefix + ' - onLeave')
  }
}

const expectOnEnter = (prefix, obj) => {
  if (typeof obj.onEnter === 'function' && obj.onEnter.mock) {
    expect(obj.onEnter.mock.calls.length).toMatchSnapshot(prefix + ' - onEnter')
  }
}

const expectThunk = (prefix, obj) => {
  if (typeof obj.thunk === 'function' && obj.thunk.mock) {
    expect(obj.thunk.mock.calls.length).toMatchSnapshot(prefix + ' - thunk')
  }
}

const expectOnComplete = (prefix, obj) => {
  if (typeof obj.onComplete === 'function' && obj.onComplete.mock) {
    expect(obj.onComplete.mock.calls.length).toMatchSnapshot(prefix + ' - onComplete')
  }
}

const expectOnError = (prefix, obj) => {
  if (typeof obj.onError === 'function' && obj.onError.mock) {
    expect(obj.onError.mock.calls.length).toMatchSnapshot(prefix + ' - onError')
  }
}


// for testing pops in a "real" browser (as simulated by Jest)

const createPop = (history) => {
  return (direction = 'back') => {
    history.currentPop = null
    window.history[direction]()
    return awaitPop(history)
  }
}

const awaitPop = async (history, tries = 1) => {
  if (tries >= 10) {
    throw new Error(`awaitPop reached the maximum amount of tries (${tries})`)
  }

  await new Promise(res => setTimeout(res, 3))
  return history.currentPop || awaitPop(history, ++tries)
}
