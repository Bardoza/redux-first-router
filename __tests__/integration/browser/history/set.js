import createTest from '../../../../__helpers__/createTest'
import { get } from '../../../../src/history/utils/sessionStorage'
import { locationToUrl } from '../../../../src/utils'
import { set } from '../../../../src/actions'


createTest('set(action)', {
  FIRST: '/:foo?'
}, {
  testBrowser: true,
  basenames: ['/base']
}, [], async ({ dispatch, getLocation }) => {
  const action = {
    query: { hell: 'yea' },
    hash: 'yolo',
    basename: '/base',
    state: { something: 123 }
  }
  await dispatch(set(action))

  expect(getLocation().hash).toEqual('yolo')
  expect(locationToUrl(window.location)).toEqual('/base/?hell=yea#yolo')
  expect(get().entries[0][0]).toEqual('/base/?hell=yea#yolo')

  expect(get()).toMatchSnapshot()
  expect(getLocation()).toMatchSnapshot()

  // for good measure, test overwriting it
  action.params = { foo: 'bar' }
  action.query = { hello: 'world', hell: undefined }
  await dispatch(set(action))

  expect(locationToUrl(window.location)).toEqual('/base/bar?hello=world#yolo')
  expect(get().entries[0][0]).toEqual('/base/bar?hello=world#yolo')

  expect(get()).toMatchSnapshot()
  expect(getLocation()).toMatchSnapshot()
})
