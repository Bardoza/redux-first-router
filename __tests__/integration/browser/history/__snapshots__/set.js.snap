// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set(action) 1`] = `
Object {
  "entries": Array [
    Array [
      "/base/?hell=yea#yolo",
      Object {
        "something": 123,
      },
      "345678",
    ],
  ],
  "index": 0,
}
`;

exports[`set(action) 2`] = `
Object {
  "basename": "base",
  "blocked": null,
  "direction": "forward",
  "entries": Array [
    Object {
      "basename": "base",
      "hash": "yolo",
      "location": Object {
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "hell=yea",
        "url": "/base/?hell=yea#yolo",
      },
      "params": Object {},
      "query": Object {
        "hell": "yea",
      },
      "state": Object {
        "something": 123,
      },
      "type": "FIRST",
    },
  ],
  "from": null,
  "hash": "yolo",
  "index": 0,
  "key": "345678",
  "kind": "load",
  "length": 1,
  "n": 1,
  "params": Object {},
  "pathname": "/",
  "pop": false,
  "prev": Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "index": -1,
      "key": "",
      "pathname": "",
      "scene": "",
      "search": "",
      "url": "",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "",
  },
  "query": Object {
    "hell": "yea",
  },
  "scene": "",
  "search": "hell=yea",
  "state": Object {
    "something": 123,
  },
  "status": 200,
  "type": "FIRST",
  "universal": false,
  "url": "/base/?hell=yea#yolo",
}
`;

exports[`set(action) 3`] = `
Object {
  "entries": Array [
    Array [
      "/base/bar?hello=world#yolo",
      Object {
        "something": 123,
      },
      "345678",
    ],
  ],
  "index": 0,
}
`;

exports[`set(action) 4`] = `
Object {
  "basename": "base",
  "blocked": null,
  "direction": "forward",
  "entries": Array [
    Object {
      "basename": "base",
      "hash": "yolo",
      "location": Object {
        "key": "345678",
        "pathname": "/bar",
        "scene": "",
        "search": "hello=world",
        "url": "/base/bar?hello=world#yolo",
      },
      "params": Object {
        "foo": "bar",
      },
      "query": Object {
        "hello": "world",
      },
      "state": Object {
        "something": 123,
      },
      "type": "FIRST",
    },
  ],
  "from": null,
  "hash": "yolo",
  "index": 0,
  "key": "345678",
  "kind": "load",
  "length": 1,
  "n": 1,
  "params": Object {
    "foo": "bar",
  },
  "pathname": "/bar",
  "pop": false,
  "prev": Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "index": -1,
      "key": "",
      "pathname": "",
      "scene": "",
      "search": "",
      "url": "",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "",
  },
  "query": Object {
    "hello": "world",
  },
  "scene": "",
  "search": "hello=world",
  "state": Object {
    "something": 123,
  },
  "status": 200,
  "type": "FIRST",
  "universal": false,
  "url": "/base/bar?hello=world#yolo",
}
`;
